Proposal: Enable purging resources in `kubectl apply`

lukemarsden opened the pull request:
Proposal to resolve kubernetes/kubernetes#19805

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29551)

<!-- Reviewable:end -->

errordeveloper committed:
Proposal for managed namespaces

lukemarsden committed:
Rework proposal a bit

lukemarsden commented:
Not sure how to add @errordeveloper to this PR. How do you add a person to a PR?

errordeveloper commented:
This CLA check is a bit confusing..

errordeveloper commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
````diff
@@ -0,0 +1,102 @@
+kubectl apply --purge-from-namespace -f https://example.com/myapp-part1.json -f https://example.com/myapp-part2.json
+```
+
+If `myapp1.json` and `myapp2.json` are in the same namespaces and it is explicitly defined, removing either of the URLs from the invocations arguments will result in all the resources that URL has defined to be purged.
````
insert `-part` here, to reflect what's above

errordeveloper commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
````diff
@@ -0,0 +1,102 @@
+
+```
+$ kubectl apply -f resources.yaml
+Updating ServiceA...
````
Currently the output doesn't even reflect whether something is being created or updated, so we will need to investigate that...

bgrant0607 replied:
That would be useful. 

The precise syntax should follow the output convention of other commands, including both the resource type and resource name in a form that could be copy+pasted to other commands verbatim, and should support `-o name` for scripting use.

bgrant0607 commented:
cc @ghodss @kubernetes/kubectl

bgrant0607 commented:
@lukemarsden @errordeveloper Commits from multiple contributors will always fail the automated CLA check.

smarterclayton commented:
Agree with general approach here - I think this is the simplest and most consistent way to solve this today.

smarterclayton commented:
We would need to warn about using the generateName field inside of your applied files.

lukemarsden commented:
> We would need to warn about using the generateName field inside of your applied files.

Yes, but we think this is a general problem with apply? We could add a warning if applied files include generateName, however. I take it that your worry is that `kubectl apply X` twice, for X with `generateName`, would end up creating two copies of the resources therein. IOW, if `generateName` creates random names, it breaks the ability for a second instantiation of `apply` to refer to the resources created by a first.

See also: https://github.com/kubernetes/kubernetes/issues/1702#issuecomment-75465841

philips commented:
cc @derekchiang @aaronlevy @ethernetdan

philips commented:
oops, I meant to cc @colhom. disregard @aaronlevy and @derekchiang

smarterclayton commented:
It is a general problem, just want to make sure that a user is aware of it
in the edge case (we don't have to fix it here, just a TODO is ok).

On Tue, Jul 26, 2016 at 12:29 PM, Brandon Philips <notifications@github.com>
wrote:

> oops, I meant to cc @colhom <https://github.com/colhom>. disregard
> @aaronlevy <https://github.com/aaronlevy> and @derekchiang
> <https://github.com/derekchiang>
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29551#issuecomment-235324606>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7W9fdIXzOtIFFLcqWJS5PMw0p-tks5qZjXngaJpZM4JUUwZ>
> .
>

sttts commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,102 @@
+
+Therefore, the proposal is to introduce an optional `--purge-from-namespace` flag to `kubectl apply`, which will allow a user to automatically purge resources from the given namespace, if they haven't been explicitly declared in files that are arguments to the same `kubectl apply` invocation.
+
+All resources referred to must be explicitly labelled with a non-default namespace in order for the `--purge-from-namespace` argument to succeed.
```
s/a non-default/one non-default/

sttts commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,102 @@
+
+All resources referred to must be explicitly labelled with a non-default namespace in order for the `--purge-from-namespace` argument to succeed.
+
+This flag will require `--namespace` not to be set.
```
I am not a big fan of hardcoding namespaces in manifests. This makes workloads less portable. E.g. in continuous deployment environments I might need to deploy/apply a workloads in a different namespace, depending on prod/testing/staging. 

Hence, my immediate thought was to enforce `--namespace` and forbid any hard-coded namespace in the manifests. This is kind of the opposite of the proposed solution.

errordeveloper replied:
Thanks, I think this is quite reasonable. The intention was that namespace must be explicit to ensure good user experience of this feature. There is wasn't a particularly big reason for doing that way, it's just that we had to pick something. I'm happy to change this over, unless @lukemarsden has something to add.

errordeveloper replied:
I've changed this.

sttts commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,102 @@
+
+This adds complexity, since a new Kubernetes API object type would need to be invented.
+
+It's not obvious how to make this concept work correctly in the case where one user runs `kubectl apply X` and later `kubectl apply X'` and a different user runs `kubectl apply Y` where `X'` is intended as an update to `X`
```
Why is that a problem? The same issue exists already with a simple `kubectl apply` command if object names are re-used. If different users have different baselines for their changes, this asks for conflicts.

errordeveloper replied:
I think @lukemarsden added this, I'm not sure what exactly was he trying to say. As this is a kind of non-critical background information, I'd vote for removing this paragraph.

lukemarsden replied:
This was in context of the alternative considered. Feel free to kill it, it was only ever a half-thought anyway.

errordeveloper committed:
Cosmetic updates

sttts commented:
Another note: there might be object kinds we want to prune and other we don't. E.g. probably endpoints (of services), events, pods of replication controllers, and probably some more.

Can we clarify this in the proposal?

sttts commented:
About my note: would a check for `generatedName != ""` be enough to know which to prune and which to keep?

errordeveloper commented:
> Another note: there might be object kinds we want to prune and other we don't. E.g. probably endpoints (of services), events, pods of replication controllers, and probably some more.

My first commit contained this line: _"There may exist a well-known annotation that can be used to exclude some resource(s) from being purged."_, but it disappeared in Luke's changes.

Would that be sufficient to address your concerns?

sttts commented:
There is a similar heuristic in the drain command (https://github.com/sttts/kubernetes/blob/739e3dd1b14743daa3368bf61f6452eb20cf7556/pkg/kubectl/cmd/drain.go#L262-L262), using the `kubernetes.io/created-by` annotation.

lukemarsden commented:
I'm fine with switching to must-specify-namespace-argument.

errordeveloper committed:
Address comments in review

errordeveloper committed:
Clarify how purgin can be disabled on some objects

errordeveloper commented:
> would a check for `generatedName != ""` be enough to know which to prune and which to keep?

I am not sure at this point. Who can clarify this? Is this the only case where we would have avoid auto purging or are there any things of similar nature?

sttts commented:
What about DaemonSet pods? Don't those have the node name attached to make
them unique?

errordeveloper commented:
@sttts #22543 is still open, so we won't be able to support DS at this point (see also #15310).

errordeveloper commented:
As this proposal is mostly about addons, and I'd expect DaemonSets to be used in addons, I think it is important to clarify whether #22543 must be a on the radar also.

bgrant0607 commented:
There are objects in kube-system that are not addons, and there will generally be resources in a namespace (e.g., secrets) that won't be managed via kubectl apply.

kubectl apply adds an annotation (`kubectl.kubernetes.io/last-applied-configuration`) to each object created/updated. Only objects with that annotation should be considered for purging. Objects without the annotation, such as pods created by ReplicationControllers, were not managed by apply, and should not be purged directly.

We could start by purging everything within a namespace, but I think we'll want to add the ability to narrow it to a specified label selector, and possibly by (list of) Kind.

smarterclayton commented:
Regarding generateName specifically, an object with a generateName value
and the apply annotation can be considered equivalent to a new object with
the same generateName field (even if the names differ).  All other cases
can be considered as deletable (for instance, multiple objects with the
same annotation and the same generated name).

On Mon, Aug 1, 2016 at 12:32 AM, Brian Grant <notifications@github.com>
wrote:

> There are objects in kube-system that are not addons, and there will
> generally be resources in a namespace (e.g., secrets) that won't be managed
> via kubectl apply.
>
> kubectl apply adds an annotation (
> kubectl.kubernetes.io/last-applied-configuration) to each object
> created/updated. Only objects with that annotation should be considered for
> purging. Objects without the annotation, such as pods created by
> ReplicationControllers, were not managed by apply, and should not be purged
> directly.
>
> We could start by purging everything within a namespace, but I think we'll
> want to add the ability to narrow it to a specified label selector, and
> possibly by (list of) Kind.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29551#issuecomment-236487150>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pwthl2lQW_Nk8Y0qcJOLzYKJROYaks5qbXbqgaJpZM4JUUwZ>
> .
>

mikedanese commented:
@smarterclayton kubectl apply doesn't work at all with objects with generateName. It blows up. I think we can solve that separately.

mikedanese commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+Currently `kubectl apply` will create any new resources that exist in the resource file(s) it is supplied with, but it has no way of deleting anything that has been removed from such files.
+
+This makes it difficult to use `kubectl apply` to update reality to match the declarative definitions of resources when that definition evolves over time, since changes can include deletions.
```
We have refereed to this process as "configuration reconciliation" in the past. https://github.com/kubernetes/kubernetes/issues/1702

mikedanese commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+This flag will require `--namespace`, and will not accept namespaces set in the input resource definitions.
+
+There shall exist a well-known annotation (`kubectl.kubernetes.io/purge-from-namespace=false`) that can be used to exclude some resource(s) from being purged.
```
I'm not sure that this is useful. This overlaps a lot with [critical addon rescheduler](https://github.com/kubernetes/kubernetes/pull/29195). We could also add a `Guarded` bit to `ObjectMeta`. This seems crude, I think we should omit it.

bgrant0607 replied:
I agree with @mikedanese

errordeveloper commented:
@bgrant0607 could you please elaborate on the suggestion about using label selectors that you've made during the call? So we have it hear in plain text for future reference.

mikedanese commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+## Proposed Solution
+
+In a "simplest possible idea first" approach, this proposal suggests that we reuse namespaces as a de-facto way of grouping resources for the purpose of _purging_ (removing) resources which no longer exist.
```
Prune seems like a more appropriate than purge for this feature.

bgrant0607 replied:
I agree with @mikedanese

mikedanese commented:
@errordeveloper The current config proposes:

> In a "simplest possible idea first" approach, this proposal suggests that we reuse namespaces as a de-facto way of grouping resources for the purpose of _purging_ (removing) resources which no longer exist.

An alternative approach to grouping is using a label selector. This has a couple advantages:

* It allows multiple configuration groups to be reconciled in the same namespace.
* By being a bit more percise, it's potentially less error prone. In a multi-user cluster and a shared namespace (thinks kube-system and a devops team), it's possible that other users have used kubectl apply to create objects in that namespace not knowing they will be wiped out.
* By convention, labels and selectors are used for grouping resources. Namespaces are used for separating quota, ACL, and names between resources. These are slightly different but selectors seem more appropriate IMO.
* Not all Kinds are namespaced. Does this feature not work with those Kinds?
* The selector language is very flexible and expressive.

The UX would change to:
````
# group all objects with a label with key kube-addon
$ kubectl apply --prune -s "kube-addon" -f addons/ 
# group all objects with a label with key kube-addon that has value heapster
$ kubectl apply --prune -s "kube-addon=heapster" -f addons/heapster
# group all objects with a label with key kube-addon that has value heapster
$ kubectl apply --prune -s "kube-addon=heapster" -f addons/heapster
# group all objects
$ kubectl apply --prune -s "" -f addons/heapster
```

The target namespace can be specified through `--namespace` or pulled from the crawled target config. Pulling from the crawled target config poses a problem. What if you delete the last object in the config directory in a namespace that was previously managed by apply? That deletion wouldn't be detected going with namespaces pulled from crawled objects.

bgrant0607 commented:
As much as possible, selectors and namespaces should work as they do in all other commands.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/get.go#L105
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/delete.go#L107
```
	cmd.Flags().StringP("selector", "l", "", "Selector (label query) to filter on")
```

Ideally we'd extend the other mutating commands to support filtering by label selector, for consistency.

Prune should additionally filter using `kubectl.kubernetes.io/last-applied-configuration`, and rely on cascading deletion of indirectly created resources, such as pods, replica sets, endpoints, etc.

mikedanese commented:
Fixed ux examples to match other commands. I agree with the points in your comment.

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+Therefore, the proposal is to introduce an optional `--purge-from-namespace` flag to `kubectl apply`, which will allow a user to automatically purge resources from the given namespace, if they haven't been explicitly declared in files that are arguments to the same `kubectl apply` invocation.
+
+All resources referred to must be explicitly labelled with one non-default namespace in order for the `--purge-from-namespace` argument to succeed.
```
I think this should state that all resources must belong to a single, non-default namespace.

"Labeled" has a specific meaning in Kubernetes, and I don't think that is what is meant here.

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+All resources referred to must be explicitly labelled with one non-default namespace in order for the `--purge-from-namespace` argument to succeed.
+
+This flag will require `--namespace`, and will not accept namespaces set in the input resource definitions.
```
I think it would be reasonable to allow the namespace from the context to be used. 

I agree that allowing other namespaces in the resource definitions could be error-prone.

Let's try to prevent accidents, but retain normal kubectl behavior where possible.

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+There exists a `/etc/kubernetes/addons` directory with some number of files. All resources declared in those files belong to the same namespace (`kube-system`), which is set explicitly in each of the resources.
+
+When `kubectl apply --purge-from-namespace -f /etc/kubernetes/addons` runs for the first time all the resources will be created.
```
This section should be updated when we agree on the syntax.

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+Subsequent invocations of `kubectl apply --purge-from-namespace -f /etc/kubernetes/addons` result in new resources being created and modified resources being updated.
+
+When user deletes some of the files or resources within files from `/etc/kubernetes/addons`, `kubectl apply --purge-from-namespace` would ensure the resources that were referenced in the deleted files are deleted, based on the assumption that `/etc/kubernetes/addons` represents all resources that are supposed to be present in `kube-system` namespace.
```
There are other resources in kube-system.

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+When user deletes some of the files or resources within files from `/etc/kubernetes/addons`, `kubectl apply --purge-from-namespace` would ensure the resources that were referenced in the deleted files are deleted, based on the assumption that `/etc/kubernetes/addons` represents all resources that are supposed to be present in `kube-system` namespace.
+
+
+## Managing Any Other Application
```
This section needs to be updated with the label-selector proposal.

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+As stated above, the `--purge-from-namespace` flag shouldn't be enabled by default, however it should be discoverable.
+
+So, if a `kubectl apply` command is run without `--purge-from-namespace`, and the referenced files are all in the same namespace, then the command could output something like:
```
I'm not sure how keen I am to do this. It will make apply significantly more expensive (to discover all available resources within the namespace) and make the user experience significantly different than other forms of updates. 

`--dry-run`, to simulate what changes would be done, would be useful.

Mentioning this in the short help snippet for kubectl apply would be useful.

A tutorial that shows how to manage resources using apply would be useful.

ghodss commented:
Our use case for apply --prune (i.e. what originally motivated #19805) is that we have a centralized git repo that has 100% of the kube objects that we want to exist in our cluster, not including objects created in response to others (e.g. pods created by replica sets). For --prune to be useful, we should be able to submit a complete list of objects, and apply should intelligently reconcile between the given list and all existing objects in the cluster *that were also originally created by apply*, including objects like namespaces. It should then delete/garbage collect all objects that are not in the list but were originally created by apply.

Given the original proposal, @mikedanese's comments, and our requirements, it sounds like the ideal design would fundamentally be based on pruning non-kubectl-apply-created objects *but* allowing for filtering with both namespace and label. The first implementation could act globally, and then both namespace and label filters could be subsequently added. What do you think?

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+    DeploymentB
+    ServiceC
+
+You can automatically delete them, if you wish, using:
```
We could suggest that on first use of apply to perform an update.

See also #19736 (suggest next commands), #10693 (more/better preferences)

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
````diff
@@ -0,0 +1,103 @@
+Done!
+```
+
+If there are > 5 resources which match, we could just list the first 5.
````
Kubectl displays all resources modified.

bgrant0607 replied:
It doesn't cover everything, but take a look at kubectl conventions:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md

bgrant0607 commented on docs/proposals/kubectl-apply-purge-from-namespace.md:
```diff
@@ -0,0 +1,103 @@
+
+If no namespace, or conflicting namespaces are provided in the resource file(s), this output would not be shown.
+
+# Alternatives Considered
```
I would just remove these.

mikedanese commented:
> be based on pruning non-kubectl-apply-created objects

I think I understand what you are saying, but to be explicit, if non-kubectl-apply-created objects are pruned it's because we cascading deleted their ancestor which *was* original kubectl-applied and is now being pruned because it doesn't exist in the target config?

> The first implementation could act globally

Is a global prune the right default? I see value in it but it seems a bit dangerous for regular usage.

bgrant0607 commented:
@mikedanese @ghodss `kubectl apply --prune` could support `--all-namespaces` for global.

ghodss commented:
> I think I understand what you are saying, but to be explicit, if non-kubectl-apply-created objects are pruned it's because we cascading deleted their ancestor which was original kubectl-applied and is now being pruned because it doesn't exist in the target config?

Yes.

> Is a global prune the right default? I see value in it but it seems a bit dangerous for regular usage.

The challenge with non-global being the default is that it's not necessarily intuitive what the default would then be. Instead we can simply require some flag to make it explicit, whether it's namespace, selector or --all.

bgrant0607 commented:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/describe.go#L101
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/get.go#L109
```
	cmd.Flags().Bool("all-namespaces", false, "If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.")
```

Note that `delete` uses `--all` to specify all resources within a namespace.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/delete.go#L108
```
	cmd.Flags().Bool("all", false, "[-all] to select all the specified resources.")
```

errordeveloper commented:
@bgrant0607 @mikedanese what is your view on how label selector would behave in cases when all of the input resource definitions are new? It also sounds like`kubectl create --selector` would have to be introduced as well, isn't it?

errordeveloper commented:
We have this in [`pkg/kubectl/cmd/delete.go#L108`][https://github.com/kubernetes/kubernetes/blob/1ac1865dd7606512e73a2687ca499049f0d97cf9/pkg/kubectl/cmd/delete.go#L108]:
```
Use:     "delete ([-f FILENAME] | TYPE [(NAME | -l label | --all)])"
```

So this means user can either provide `-l` or `-f`, but not both. With `-l` deletion would be performed on the objects returned by the selector, while with `f` it would be performed on what's defined in the input files. With this in mind, I am not sure how to interpret `-l` and `-f` used along side with examples proposed above. Also, if we allow `-f` and `-l` in the case of `kubectl apply`, should it also be reflect in `kubectl delete`, and how would people see that working exactly?

bgrant0607 commented:
@errordeveloper Yes, we should extend existing commands, including get, create, and delete, to be able to filter the set of objects processed using a label selector, whether those objects are identified by files/URLs/stdin, by name, by listing resources within a namespace, etc.

lukemarsden commented:
@bgrant0607 @mikedanese 

I can see how pruning could work given a label selector, if indeed the user is using labels to organize their objects, which seems reasonable. And we could certainly start managing objects for add-ons in the context of turnup. (Also: ack, we should only prune objects with annotation `kubectl.kubernetes.io/last-applied-configuration`.)

However, where I get stuck is on trying to understand what labels (if any) should be applied to objects _created_ by a `kubectl apply -l $label_selector -f resources.yaml`.

If `resources.yaml` doesn't specify any labels, should the labels in `$label_selector` be applied to the created resources?

If yes, then – well, erk: according to my read of the [label selectors](http://kubernetes.io/docs/user-guide/labels/#label-selectors) doc, label selectors are a conjunction of equality- or set-based conditions. It's non-trivial to go from an arbitrary label selector to a set of labels that should be applied to the created objects.

If no, then – well, the resulting system doesn't satisfy the criteria that:

* `kubectl apply -l $label_selector -f resources.v1.yaml`
[delete an object in resources.v1.yaml, rename it resources.v2.yaml]
* `kubectl apply -l $label_selector -f resources.v2.yaml`

is guaranteed to delete objects that were removed between `v1` and `v2`. It would only work if the user had remembered to label the objects in `resources.v1.yaml` in a way which matches with the label selector.

I might be missing something here, what would you advise?

Thanks!

lukemarsden commented:
BTW, it might be that "no" is OK. In which case I can update the proposal and we'll just have to be sure to carefully documented the expected usage of this command _and_ expected conventions for labels.

mikedanese commented:
I think the answer is no. Is the concern that the interaction between -l and the labels in an applied object set will be too complex for the average user?

bgrant0607 commented:
@lukemarsden:

@mikedanese is correct that the answer is "no".

In general, Kubernetes expects users to add identifying labels to their objects, and changing identifying labels, like changing object names, is not going to end well for the user. That's not specific to apply.

lukemarsden commented:
Thanks @mikedanese and @bgrant0607, we will update the proposal and start coding this Soon™.

lukemarsden committed:
Simplify proposal and update it to use labels, not namespaces.

lukemarsden committed:
Longer but clearer name.

lukemarsden committed:
Rename file

lukemarsden committed:
Clarify behavior.

lukemarsden commented:
@mikedanese @bgrant0607 

I've simplified the proposal and updated it to reflect the discussion above. Please take another look: https://github.com/errordeveloper/kubernetes/blob/kubectl-apply-purge-from-namespace/docs/proposals/kubectl-apply-purge-missing-where.md

Thanks!

ghodss commented:
This is very cool, thanks for the update! Couple questions:

1) How does this interact with namespaces? Maybe this was implicit but here's how I would imagine it:

1. If no namespace is specified, operate across `default` namespace only.
1. If `--namespace` is specified, only operate in that namespace.
1. If `--all-namespaces` is specified, operate across the entire cluster.

...with the label selector working as expected in all cases. I do think you could make a case that if you do `--all-namespaces`, adding a label selector should throw an error, since labels generally are designed to be scoped within namespaces. But I could be convinced either way on that.

2) If you just do `--purge-missing-where ""`, do you then affect all objects? This makes sense to me, but I wonder if there should be a different option or change the option name for it to be more intuitive in both cases.

errordeveloper commented:
@ghodss good question indeed, sounds like this should be clarified.

@lukemarsden I think the title of this PR no longer reflects the current revision of the proposal doc, doesn't it? It's really not so much about purging from namespaces any more.

lukemarsden commented:
@ghodss thanks for the review!

1. I hadn't thought about this, but I concur with your reasoning. I will update the proposal with a section on namespace interactions, cnp'ing your suggestion. Personally I wouldn't add the `--all-namespaces` implies disabling the flag logic, because it makes the system less flexible. Users should be able to label things across namespaces if they wish, even if it's not the recommended configuration.

2. `--purge-missing-where ""` is a dangerous operation, I'm OK with it being a bit hidden/cryptic. Just as long as `--purge-missing-where` without the `""` doesn't have the same effect. But maybe it could print a warning which can only be overridden with `--force`? WDYT?

@bgrant0607 not urgent, but do you have any comments on the above?

lukemarsden commented:
@errordeveloper title fixed, thanks!

lukemarsden committed:
Specify namespace interactions, thanks to @ghodss

lukemarsden commented:
Updated the proposal with namespace bits, thanks @ghodss!

https://github.com/kubernetes/kubernetes/pull/29551/files#diff-a269fcb268448f592cf99796909b24a6R48

bgrant0607 commented on docs/proposals/kubectl-apply-purge-missing-where.md:
```diff
@@ -0,0 +1,54 @@
+
+## Interactions with namespaces
+
+* If no namespace is specified, operate across `default` namespace only.
```
Namespace behavior should be the same as all other kubectl commands, which means that namespace comes from the context by default.

bgrant0607 commented:
`delete` has a `--all` flag to specify all resources in the namespace.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/delete.go#L108

bgrant0607 commented on docs/proposals/kubectl-apply-purge-missing-where.md:
````diff
@@ -0,0 +1,54 @@
+The usage may look like this (assuming the user uses the convention of labelling apps with the `app` label):
+
+```
+kubectl apply --purge-missing-where 'app=myapp' -f https://example.com/myapp.json
````
Why mash `--purge` and the label selector together into a single flag? We have a convention for filtering commands by label selector already.

lukemarsden replied:
@bgrant0607 because the label selector is _specific to the purge operation_. Splitting the purge flag from the label selector, IMO, makes it more dangerous, because presumably --purge without a label selector would delete everything not in the provided manifests. Making --purge require -l would beg the question why not unify them? Furthermore the spelling of --purge-missing-where "condition" reads naturally to explain the semantics of what's happening IMO. --purge -l condition is less clear to new users. Furthermore -l makes sense as a filter on other ops because they are simple ops, like delete - it's obvious what's being filtered. But we're making apply a compound operation: "create some things and delete others", -l on its own isn't sufficiently descriptive to make clear what the label selector applies to (only the deletions). That's why I proposed the combined flag -- it's better UX IMO.

ghodss replied:
With the original design which was entirely label based, merging the flags certainly had merit. But now that we've introduced namespace as a filter as well, doesn't it break down a little? The filter is by selector, namespace, both or none, so at this point, having the purge flag merged specifically with the label selector doesn't seem to make as much sense.

If the goal is to protect against the danger of the flag, we should add a confirmation step which by default lists all resources which will be deleted and requires user confirmation to continue. Then we can add another flag like `--continue` to skip the prompt.

That way we protect against user error while maintaining our consistency of `-l` and `--namespace`.

bgrant0607 replied:
@lukemarsden I disagree that the selector should only apply to prune/purge. I'd like to be able to use it to update a subset of the objects. 

Also, we need to consider the overall usability of kubectl, not just this command. I prioritize consistency across commands fairly highly, so there's a high bar for departing from established precedent.

smarterclayton replied:
That pattern (confirmation for dangerous actions) is well enshrined
via "--confirm" in other places.

bgrant0607 commented:
@lukemarsden Please do not deviate from conventions established by other commands.

bgrant0607 commented:
@lukemarsden Please re-read prior feedback.

lukemarsden commented on docs/proposals/kubectl-apply-purge-missing-where.md:
```diff
@@ -0,0 +1,54 @@
+## Interactions with namespaces
+
+* If no namespace is specified, operate across `default` namespace only.
+* If `--namespace` is specified, only operate in that namespace.
```
. Z.F h C

lukemarsden commented:
@bgrant0607 @ghodss @smarterclayton thanks for the feedback! I'll address this in another rev of the proposal shortly.

bgrant0607 commented:
@lukemarsden Thanks.

Another way to look at this: Assume that `apply` will get some of the same flags, such as `--selector`, as other commands (e.g., `delete`, `get`, `replace`).  How should `--purge-missing` interact with those flags?

bgrant0607 commented:
Relevant flags:
* `--selector`
* `--all`
* `--all-namespaces`

I assume that by default the deletion behavior will correspond to `delete --cascade=true --ignore-not-found=true --now=false`.

bgrant0607 commented:
@lukemarsden Is this proposal still active? I would like to see it finished.

mikedanese commented:
FYI, there is a WIP implementation here https://github.com/kubernetes/kubernetes/pull/33075

bgrant0607 commented:
#33075 merged.
